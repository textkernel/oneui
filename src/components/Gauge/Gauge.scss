@import '../../themes/oneui/oneui.scss';

.Gauge {
    position: relative;

    &__svg {
        transform: rotate(180deg);
    }

    &__circleBackground {
        stroke: var(--color-neutral-25);
        stroke-width: var(--spacing-2x);
        stroke-linecap: round;
        fill: none;
    }

    &__circleForeground {
        stroke-width: var(--spacing-4x);
        stroke-linecap: round;
        fill: none;
        transition: all var(--transition-duration) ease-out;

        &--isProgressLoading {
            opacity: 0;
        }
    }

    &__contentPlaceholder {
        margin-bottom: 0;
    }

    &__contentWrapper {
        position: absolute;
        top: 50%;
        left: 50%;
        margin-top: var(--spacing-normal);
        transform: translate(-50%, -100%);
        text-align: center;

        &--isContentLoading {
            width: 100px;
        }
    }

    &__content {
        display: inline;
        font-size: var(--h1-font-size);
        font-weight: var(--font-weight-bold);
    }

    &__metric {
        display: block;
        width: 150px;
        font-size: var(--font-size-normal);
        font-weight: var(--font-weight-normal);
        white-space: nowrap;
        text-overflow: ellipsis;
        overflow: hidden;
    }

    &__note {
        position: absolute;
        top: 50%;
        left: 50%;
        transform: translate(-50%, 100%);
        color: var(--color-muted);
        text-transform: uppercase;
        font-weight: var(--font-weight-bold);
        font-size: var(--font-size-small);
    }

    @each $context in ($contexts) {
        &__gradientStart {
            &--context_#{ $context } {
                stop-color: var(--color-#{$context});
            }
        }

        &__gradientEnd {
            &--context_#{ $context } {
                stop-color: var(--color-#{$context}-25);
            }
        }
    }
}
