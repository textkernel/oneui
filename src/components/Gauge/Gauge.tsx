import * as React from 'react';
import { bem } from '../../utils';
import { ContentPlaceholder } from '../ContentPlaceholder';
import { GAUGE_RADIUS } from '../../constants';
import styles from './Gauge.scss';

const SET_PERCENTAGE_DELAY = 100;

const { block, elem } = bem('Gauge', styles);

export interface Props {
    /** Defines if progress bar is in loading state */
    isProgressLoading?: boolean;
    /** Defines if content part is in loading state */
    isContentLoading?: boolean;
    /** Percentage of the progress bar to be filled */
    percentage?: number;
    /** Content component */
    children: React.ReactNode;
    /** String value to be used as metric part below the child */
    metric?: string;
    /** Note component to be rendered below the gauge */
    note?: React.ReactNode;
}

/**
 * ## Testing information
 * This component is uses `Math.random` internally which will show up in snapshots.
 * To make sure snapshots are consistent you should mock it in your tests. E.g.:

      const mathRandomSpy = jest.spyOn(Math, 'random');
      mathRandomSpy.mockImplementation(() => 0.42);
 */
export const Gauge: React.FC<Props> = ({
    children,
    percentage = 0,
    note,
    metric = '',
    isProgressLoading = false,
    isContentLoading = false,
    ...rest
}) => {
    const [percentageToShow, setPercentageToShow] = React.useState(0);
    // We need this uuid for the linear gradient's id to avoid some bugs in Safari.
    // For details see: https://jira.textkernel.nl/browse/JF-2780

    // Simulate a prop change to run the animation on render.
    // There's a small delay set because of a weird bug
    // with preventing of the animation being run with a child wrapped up into a Tooltip.
    React.useEffect(() => {
        const percentageTimeout = setTimeout(() => {
            setPercentageToShow(percentage);
        }, SET_PERCENTAGE_DELAY);

        return () => {
            clearTimeout(percentageTimeout);
        };
    }, [percentage]);

    const percentageAdjusted = Math.max(0, Math.min(100, percentageToShow)) / 100;
    const progress = isProgressLoading ? 0 : percentageAdjusted;
    const circumference = 2 * Math.PI * GAUGE_RADIUS;
    const circumferenceHalf = circumference / 2;

    if (typeof children !== 'number' && !children) {
        return null;
    }

    return (
        <div {...rest} {...block({ isProgressLoading, isContentLoading, ...rest })}>
            <svg {...elem('svg')} width="100%">
                <circle
                    {...elem('circleBackground')}
                    r={GAUGE_RADIUS}
                    cx="50%"
                    cy="25%"
                    style={{ strokeDasharray: `${circumferenceHalf} ${circumference}` }}
                />
                <circle
                    {...elem('circleForeground', { isProgressLoading })}
                    r={GAUGE_RADIUS}
                    cx="50%"
                    cy="25%"
                    style={{ strokeDasharray: `${progress * circumferenceHalf} ${circumference}` }}
                />
            </svg>
            <div {...elem('contentWrapper', { isContentLoading })}>
                {isContentLoading ? (
                    <ContentPlaceholder height={28} {...elem('contentPlaceholder')} />
                ) : (
                    <div {...elem('content')}>
                        {children}
                        {metric ? (
                            <div {...elem('metric')} title={metric}>
                                {metric}
                            </div>
                        ) : null}
                    </div>
                )}
            </div>
            {note ? <div {...elem('note')}>{note}</div> : null}
        </div>
    );
};

Gauge.displayName = 'Gauge';
