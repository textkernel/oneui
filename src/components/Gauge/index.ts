export { Gauge, Props as GaugeProps } from './Gauge';
