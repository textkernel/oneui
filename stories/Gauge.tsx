/* eslint-disable react/no-children-prop */
import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { boolean, number, text, select, withKnobs } from '@storybook/addon-knobs';
import { Gauge, IconJobfeed, Tooltip } from '@textkernel/oneui';
import { CONTEXTS } from '@textkernel/oneui/constants';

storiesOf('Molecules|Gauge', module)
    .addDecorator(withKnobs)
    .add(
        'Gauge',
        () => {
            const isProgressLoading = boolean('isProgressLoading', false);
            const isContentLoading = boolean('isContentLoading', false);
            const note = text('Note', 'Low');

            return (
                <div style={{ display: 'flex' }}>
                    <div style={{ width: '33%' }}>
                        <Gauge
                            context={select('Context 1', CONTEXTS, CONTEXTS[1])}
                            percentage={number('Percentage 1', 33)}
                            note={isContentLoading ? '—' : note}
                            metric={text('Metric', 'Postings per job')}
                            isProgressLoading={isProgressLoading}
                            isContentLoading={isContentLoading}
                        >
                            {text('Children', '3')}
                        </Gauge>
                    </div>
                    <div style={{ width: '33%' }}>
                        <Gauge
                            context={select('Context 2', CONTEXTS, CONTEXTS[1])}
                            percentage={number('Percentage 2', 46)}
                            note={isContentLoading ? '—' : 'Normal'}
                            metric="Days"
                            isProgressLoading={isProgressLoading}
                            isContentLoading={isContentLoading}
                        >
                            12
                        </Gauge>
                    </div>
                    <div style={{ width: '33%' }}>
                        <Gauge
                            context={select('Context 3', CONTEXTS, CONTEXTS[2])}
                            percentage={number('Percentage 3', 78)}
                            note={isContentLoading ? '—' : 'High'}
                            isProgressLoading={isProgressLoading}
                            isContentLoading={isContentLoading}
                        >
                            $54,321
                        </Gauge>
                    </div>
                </div>
            );
        },
        {
            info: {
                text: `
            ## Testing information

            This component is uses \`Math.random\` internally which will show up in snapshots.

            To make sure snapshots are consistent you should mock it in your tests. E.g.:

            \`\`\`
            const mathRandomSpy = jest.spyOn(Math, 'random');
            mathRandomSpy.mockImplementation(() => 0.42);
            \`\`\`

            `,
            },
        }
    )
    .add('Gauge with an icon', () => {
        const isProgressLoading = boolean('isProgressLoading', false);
        const isContentLoading = boolean('isContentLoading', false);

        return (
            <div style={{ display: 'flex' }}>
                <Gauge
                    context="primary"
                    percentage={60}
                    note={isContentLoading ? '—' : 'High'}
                    isProgressLoading={isProgressLoading}
                    isContentLoading={isContentLoading}
                >
                    <div>
                        $54,321
                        <IconJobfeed context="brand" />
                    </div>
                </Gauge>
            </div>
        );
    })
    .add('Gauge with a tooltip', () => {
        const isProgressLoading = boolean('isProgressLoading', false);
        const isContentLoading = boolean('isContentLoading', false);

        return (
            <div style={{ display: 'flex' }}>
                <Gauge
                    context="primary"
                    percentage={35}
                    metric={isContentLoading ? '—' : "Something long so it doesn't fit"}
                    note={isContentLoading ? '—' : 'High'}
                    isProgressLoading={isProgressLoading}
                    isContentLoading={isContentLoading}
                >
                    <Tooltip content="Tooltip content">
                        <span>Tooltip</span>
                    </Tooltip>
                </Gauge>
            </div>
        );
    });
