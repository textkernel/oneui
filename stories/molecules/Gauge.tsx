import * as React from 'react';
import { ComponentStory } from '@storybook/react';
import { Gauge, IconJobfeed, Tooltip } from '@textkernel/oneui';

export default {
    title: 'Molecules/Gauge',
    component: Gauge,
};

const Template: ComponentStory<typeof Gauge> = (args) => (
    <div style={{ display: 'flex' }}>
        <Gauge {...args} />
    </div>
);

export const _Gauge = (args) => {
    return (
        <div style={{ display: 'flex' }}>
            <div style={{ width: '33%' }}>
                <Gauge {...args} />
            </div>
            <div style={{ width: '33%' }}>
                <Gauge {...args}>12</Gauge>
            </div>
            <div style={{ width: '33%' }}>
                <Gauge {...args}>$54,321</Gauge>
            </div>
        </div>
    );
};
_Gauge.args = {
    percentage: 33,
    metric: 'Postings per job',
    children: '3',
    note: 'Low',
};

export const GaugeWithAnIcon = Template.bind({});
GaugeWithAnIcon.storyName = 'Gauge with an icon';
GaugeWithAnIcon.args = {
    percentage: 60,
    note: 'High',
    isProgressLoading: false,
    isContentLoading: false,
    children: (
        <div>
            $54,321
            <IconJobfeed />
        </div>
    ),
};

export const GaugeWithATooltip = Template.bind({});
GaugeWithATooltip.storyName = 'Gauge with a tooltip';
GaugeWithATooltip.args = {
    percentage: 35,
    metric: "Something long so it doesn't fit",
    note: 'High',
    isProgressLoading: false,
    isContentLoading: false,
    children: (
        <Tooltip content="Tooltip content">
            <span>Tooltip</span>
        </Tooltip>
    ),
};
