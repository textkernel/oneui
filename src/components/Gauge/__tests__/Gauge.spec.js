import React from 'react';
import toJson from 'enzyme-to-json';
import { Gauge } from '../Gauge';

describe('Gauge', () => {
    const mathRandomSpy = jest.spyOn(Math, 'random');
    mathRandomSpy.mockImplementation(() => 0.42);

    it('should render nothing if no children are provided', () => {
        const wrapper = shallow(<Gauge />);
        expect(wrapper.html()).toBeNull();
    });
    it('should render correctly', () => {
        const wrapper = shallow(
            <Gauge percentage={33} note="Value" metric="Metric">
                Some content
            </Gauge>
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render with loading state correctly', () => {
        const wrapper = shallow(
            <Gauge metric="Metric" isContentLoading isProgressLoading>
                Some content
            </Gauge>
        );
        expect(wrapper.find('ContentPlaceholder').exists()).toBeTruthy();
    });
});
