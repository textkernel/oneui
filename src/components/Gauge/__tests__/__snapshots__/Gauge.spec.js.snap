// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gauge should render correctly 1`] = `
<div
  className="Gauge"
>
  <svg
    className="Gauge__svg"
    width="100%"
  >
    <defs>
      <linearGradient
        id="Gauge__gradient--brand-0.42"
      >
        <stop
          className="Gauge__gradientStart--context_brand"
          offset="50%"
        />
        <stop
          className="Gauge__gradientEnd--context_brand"
          offset="100%"
        />
      </linearGradient>
    </defs>
    <circle
      className="Gauge__circleBackground"
      cx="50%"
      cy="25%"
      r={100}
      style={
        Object {
          "strokeDasharray": "314.1592653589793 628.3185307179587",
        }
      }
    />
    <circle
      className="Gauge__circleForeground"
      cx="50%"
      cy="25%"
      r={100}
      stroke="url(#Gauge__gradient--brand-0.42)"
      style={
        Object {
          "strokeDasharray": "0 628.3185307179587",
        }
      }
    />
  </svg>
  <div
    className="Gauge__contentWrapper"
  >
    <div
      className="Gauge__content"
    >
      Some content
      <div
        className="Gauge__metric"
        title="Metric"
      >
        Metric
      </div>
    </div>
  </div>
  <div
    className="Gauge__note"
  >
    Value
  </div>
</div>
`;
